handle omission of opening quote mark

davideleuterius opened the pull request:
This should address omission of the opening quote marks on drop caps.

davideleuterius committed:
handle omission of opening quote mark

designerbrent reviewed (changes requested):
While this handles quote marks, it doesn't deal with HTML well, as you can see. 

![2020-06-17 at 4 07 PM](https://user-images.githubusercontent.com/132864/84945033-baa1f700-b0b4-11ea-9963-5ed07e957913.png)

Due to the complexities, I'm actually wondering if we can keep the same markup we have now, the `has-dropcap` class, but use `::first-letter` structure that Wordpress uses. This is my hackish attempt at making it work strictly in the browse. Yes, this example does have a quote inside, but that's not the worst thing.
![2020-06-17 at 4 09 PM](https://user-images.githubusercontent.com/132864/84945222-05237380-b0b5-11ea-98c2-6fa2df3bb554.png)

In doing this, we have to be able to handle: 

- Normal text
- Quoted text
- <strong> or <em> text

davideleuterius commented:
@designerbrent That sounds like a good approach. I'll just use `first-letter` and that should handle all those possible scenarios. I'll have that up shortly.

designerbrent commented:
@davideleuterius How are things coming on this issue?

davideleuterius commented:
@designerbrent - sorry I have been delayed on this. I'm in the middle of working this up, which turned out to be a little more involved than you would think due to dealing with the possibility of an html tag starting the content, so I'm having to look at using a regular expression to remove that tag for applying the style. 

I'll be getting back on this tomorrow or Thursday as I've been making a big lift to make a deadline for another project, but I can then return to this and get a solution to you. Thanks for your patience as I work through this.

designerbrent commented:
@davideleuterius Any word on this?

davideleuterius commented:
Brent,

My apologies. I got sucked into making some immediate updates after the deadline. I will work on this immediately to bring this to a close now.

davideleuterius committed:
drop cap scripting

- handles opening quote
- detects opening tag, splits inner text, removes first letter & preserves tag
- sets first letter as drop cap

davideleuterius commented:
@designerbrent 
Got this done for you. The complexity here was preserving the tag when it is the first element that a drop cap is being applied to, so rather than remove the tag, we're finding the inner text on it, splitting it from the first letter, setting the inner text to the split string (minus the first letter), and setting the first letter in a new html container where the drop cap style can be applied. I have tested this with opening quotes, the strong tag (although any tag should work), and no opening quote or tag. It applied the drop cap successfully for me in each case.

Let me know if this works for you or if you find an exception that needs to be addressed here, and thanks for your patience!

designerbrent reviewed (changes requested):
@davideleuterius I tried testing this and couldn't find a way to make it even show up. I looked through the code and found you have `o-dropcap` and `o-drop-cap` classes as well as some similar variations in the Javascript code. 

Is there a reason we have two different conventions? 

Can you please make a demo page for this pattern that includes paragraphs with no html, starting with tags, and starting with quotes?

davideleuterius committed:
adding dev / test page for drop caps

davideleuterius commented:
@designerbrent 
On the WordPress theme, when I have the drop cap toggled in the text settings for the block, it is applying a class of .has-drop-cap to the paragraph: 

![drop-cap-from-wp](https://user-images.githubusercontent.com/4670424/87174346-a492eb00-c29c-11ea-82d6-bafee5ec0eca.png)

So this is what I have targeted with the scripting. The scripting is looking to transform the following to a drop cap:

'p.has-drop-cap, .has-drop-cap p:eq(0)'

This will select any paragraph that has the drop cap class, or the first paragraph in any parent container that has the class. This class is solely for targeting for this scripting, following how it is set in WP currently. In WP it seems to set it exclusively on the P tag, but if this intended for use elsewhere, I had thought it might be possible that the drop cap could be applied to a parent container if outside of the WP theme.

Regarding the different classes for the drop cap, the issue here is that there is an existing class of `o-dropcap` that uses the first-letter / data-letter attribute method that will not work with leading tags or quotes. Rather than delete this and perhaps cause issues with existing content and code, I created a new `o-drop-cap` class ( similarto the WP naming convention for consistency ) that will construct the drop cap styling working with our new methods. The old class cannot build the drop cap successfully moving forward, so I created this new class to recreate the same presentation and handle the new method for building it.

I'll push up a test page I was using for this. Let me know if this makes sense or what your thoughts are on this for bringing it in alignment with your expectations - thanks!

designerbrent commented:
@davideleuterius Sorry I didn't get back to you sooner on this. I think it makes sense what you are saying, but I'm struggling to test it. I'm wondering if you could sync this branch up with the `v3.x` branch as that is using a newer version of `grunt` and `npm` and I am able to run that.

This will need to happen to merge this anyway to merge it.

davideleuterius committed:
Merge branch 'v3.x' into v3.459-drop-caps

davideleuterius commented:
@designerbrent 
I just merged the current v3.x branch into this PR. There should be a 'dropcap' pattern to view for you now, but it isn't anything special at all. It was a quick simple test to see how the dropcap would be applied. Let me know if this works for you or if there is anything else I can do. I've got some time this afternoon to get this shored up for you - thanks!

designerbrent committed:
Adding some sizing/spacing adjustments and more test cases. Also renamed all the "has-dropcap" instances to "has-drop-cap" in the templates.

designerbrent committed:
Fixing the dropcap for RTL.

designerbrent commented on source/_patterns/01-molecules/text/dropcap.twig:
```diff
@@ -0,0 +1,45 @@
+<section class="l-grid l-grid--7-col u-shift--left--1-col--at-xxlarge l-grid-wrap--6-of-7 u-spacing--double--until-large  u-space--quad--top ">
+  <div class="c-article l-grid-item l-grid-item--l--4-col l-grid-item--xl--3-col">
+    <article class="text c-article__body u-spacing--double has-drop-cap">
+      <p><strong>Lorem ipsum dolor sit amet,</strong> consectetur adipiscing elit. <a href="">Donec malesuada</a>, est ut viverra euismod, dui dolor
```
@davideleuterius I added another testing block that mimics what happens on an article. However I can't get the dropcap to apply here, though it does work on the main article pages now. Can you help me understand why it doesn't work on this example?

davideleuterius replied:
@designerbrent 
I'll take a look at this today to see why this isn't applying as desired here.

designerbrent replied:
This is working correctly now.

designerbrent commented on source/js/script.js:
```diff
@@ -134,18 +134,28 @@
     return el.childNodes[0];
   }
 
-  // Grab first character for dropcaps
```
Removing this will affect all other themes that use this old method. I'm wondering if we should leave this in longer to enabled the ModX theme to make this change?

davideleuterius replied:
@designerbrent 
I'll review to see if leaving this in causes any conflicts. If not, then that makes sense - no need to remove it if it would be a breaking change for some cases.

designerbrent replied:
Any word on this, @davideleuterius?

designerbrent reviewed (changes requested):
I've made a few adjustments to the CSS, as well as the existing patterns to make use the new class. Short the two questions above, I think this is good.

designerbrent commented:
@claytonk @eduludi This change will affect your versions of ALPS, with the removal of the `has-dropcap` (in ModX) and the future removal of `o-dropcap`. These will be replaced with `has-drop-cap` and `o-drop-cap` respectively.

designerbrent commented:
@davideleuterius Once this gets committed, we need to remove the `has-dropcap` functions in Wordpress: [the results here.](https://github.com/adventistchurch/alps-wordpress/search?q=dropcap&unscoped_q=dropcap)

davideleuterius commented:
@designerbrent - thanks for pointing that out. I'll be sure to remove that function in WordPress as it will no longer serve a purpose.

davideleuterius committed:
adding specific checks - dropCapChecks

davideleuterius commented:
@designerbrent 
I made a more specific / explicit call for finding / setting the drop cap and it seems to work. I am not sure why the more generic call didn't set it, but left it just in case, and followed it with the more specific calls to catch this use case. Let me know if you see anything else needing attention and thanks!

designerbrent committed:
Adding some descriptions to the test case.

davideleuterius committed:
Merge branch 'v3.x' into v3.459-drop-caps

davideleuterius committed:
merge & restoring legacy drop cap implementation

davideleuterius commented:
@designerbrent 
I merged the 3.x branch and also restored the older / legacy function for drop caps. Let me know if this looks right or if there's more work to do here - thanks!

designerbrent commented:
Thanks @davideleuterius. I'm going to sit on this till next week while we prep to make some changes on 2 other themes. Would you be able to get me a PR for the `ALPS for Wordpress` theme by Thursday that removes the Dropcap checkbox that we added with Carbon Fields?

davideleuterius commented:
@designerbrent 
I can. Also how should that be handled in the Guteberg block? There's a drop cap toggle for the paragraph in one of the blocks. Is the expectation that the first letter of a page/post will always have the drop cap now?

designerbrent commented:
@davideleuterius 
Good questions. If we can do something to default the first Gutenberg block to have that turned on, that would be nice, but I don't think we need to go to heroics to make that happen. I don't want a custom paragraph block.

davideleuterius commented:
@designerbrent 
Going to get this to you for WP now. Can you create an issue (or point me to an existing one) for this so I can name the branch right for the drop cap checkbox removal in WP? Or I can do it if you prefer, but I am not sure I have permissions to do so? Thanks.

davideleuterius commented:
@designerbrent 
I have this ready to go - maybe I should just push this up as v3.drop-cap-checkbox? Seeing as how there isn't an issue assigned for this?

designerbrent commented:
@davideleuterius I just created the issue here: https://github.com/adventistchurch/alps-wordpress/issues/508

designerbrent committed:
Merge branch 'v3.x' into v3.459-drop-caps
